### PR TITLE
[stable/influxdb] Adds support for InfluxDB Enterprise

### LINKED ISSUE #18685: [stable/influxdb] PR for adding optional support of enterprise
@jackzampolin 
@aisuko

Would you be interested in a PR for adding in support for InfluxDB Enterprise and replications, perhaps building off of #17619?

**Is your feature request related to a problem? Please describe.**
I plan on writing a chart for this anyways, so it would likely make sense to build off of the existing chart, and be able to contribute it back. 

**Describe the solution you'd like**
Adding in conditional support for meta servers, license configmap, and replication (probably other things)

**Describe alternatives you've considered**
Making a chart based on https://docs.influxdata.com/enterprise_influxdb/v1.7/install-and-deploy/production_installation/meta_node_installation/

### PR DESCRIPTION
**Pull request by wgj**
Follows the newly introduced pattern of StatefulSets for pods, extending InfluxDB's
sts and allowing parameter for it's ReplicaSet, and adds a new sts for InfluxDB
"meta nodes" which hold distribution state for InfluxDB's high availability and
horizontal scaling.

Documentation for these new features has been added, but I can not stress enough
how important it is to follow, as Enterprise not only uses different container
images for meta containers, but also expects new container images for InfluxDB
(appropriately labeled `data` in Docker Hub) which in my experience was easy to
miss if a user is following one set of [instructions](https://docs.influxdata.com/enterprise_influxdb/v1.7/install-and-deploy/production_installation/data_node_installation/) over [another](https://hub.docker.com/_/influxdb).

A caveat of trade off of deploying Enterprise is an expectation of human
interaction when creating or scaling the cluster (`influxd-ctl` usage). This
makes Enterprise decidedly not "cloud native", but is less obnoxious when using
a StatefulSet with PVCs, and no pod auto-scaling, of which I suppose is unlikely
given InfluxData's approach to licensing.

Still forcing k8s operator involvement feels "bad", and I left in provisions to
improve this expectation in a later version, namely a headless service and using
`publishNotReadyAddresses`, such that a pod can use the Service to find
available `influxdb-meta` and `influxdb` pods to create/join an InfluxDB
cluster.

Signed-off-by: Weston Johnson <wgj@automox.com>

#### Is this a new chart
No.

#### What this PR does / why we need it:
Adds support for InfluxDB Enterprise, specifically HA clustering.

#### Which issue this PR fixes
  - fixes #18685

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by Aisuko** [comment]
Thanks a lot, the new feature is so awesome. @wgj Please solute the conflict.  cc @naseemkullah Hello, sir. If you can be here help us to review or give some advice, it is our pleasure.

File: stable/influxdb/templates/statefulset.yaml [hunk 1]
```diff
@@ -58,9 +62,19 @@ spec:
...
+        - name: rpc
+          containerPort: {{ .Values.config.rpc.bind_address }}
+        - name: meta
+          containerPort: {{ .Values.config.meta.bind_address }}
+        {{- end }}
         {{- if .Values.env }}
         env:
 {{ toYaml .Values.env | indent 10 }}
+        {{- if .Values.enterprise.enabled }}
+          - name: INFLUXDB_HOSTNAME # Values.env's HOSTNAME isn't fundamentally different from $HOSTNAME, but this way weg get a distinguished name at runtime.
```
Comment (naseemkullah) [inline stable/influxdb/templates/statefulset.yaml:92]: Not familiar with enterprise so not sure how/what this is used for, but would `"{{ template "influxdb.fullname" . }}-<stateful set replica number>` (Which is value for $HOSTNAME) not be sufficient?

File: stable/influxdb/templates/meta-statefulset.yaml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        - name: rpc
+          containerPort: {{ .Values.config.rpc.bind_address }}
+        - name: meta
+          containerPort: {{ .Values.config.meta.bind_address }}
+        {{- if .Values.env }}
+        env:
+{{ toYaml .Values.env | indent 10 }}
+          # Values.env's HOSTNAME isn't fundamentally different from $HOSTNAME, but this way we get a distinguished name for InfluxDB at runtime.
+          - name: INFLUXDB_HOSTNAME
+            value: "$(_HOSTNAME).{{ template "influxdb.fullname" . }}-meta"
```
Comment (naseemkullah) [inline stable/influxdb/templates/meta-statefulset.yaml:51]: Not familiar with enterprise so not sure how/what this is used for, but would `"{{ template "influxdb.fullname" . }}-meta-<stateful set replica number>` (Which is value for $HOSTNAME) not be sufficiently distinguished?

File: stable/influxdb/values.yaml [hunk 3]
```diff
@@ -56,6 +56,29 @@ persistence:
   accessMode: ReadWriteOnce
   size: 8Gi
 
+## Deploy InfluxDB Enterprise - License required
+##
+enterprise:
+  enabled: false
+  licensekey: {}
```
Comment (naseemkullah) [inline stable/influxdb/values.yaml:63]: Would be nice to add an option to provide the license key via a pre existing k8s secret

File: stable/influxdb/templates/meta-statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            storage: {{ .Values.persistence.size | quote }}
+      {{- if .Values.persistence.storageClass }}
+      {{- if (eq "-" .Values.persistence.storageClass) }}
+        storageClassName: ""
+      {{- else }}
+        storageClassName: "{{ .Values.persistence.storageClass }}"
+      {{- end }}
+      {{- end }}
+  {{- end }}
+{{- end }}
```
Comment (naseemkullah) [inline stable/influxdb/templates/meta-statefulset.yaml]: nit: empty line at EOF

File: stable/influxdb/README.md [hunk 5]
```diff
@@ -65,6 +65,39 @@ $ helm install --name my-release -f values.yaml stable/influxdb
 
 > **Tip**: You can use the default [values.yaml](values.yaml)
 
+### Enterprise
+[InfluxDB Enterprise](https://www.influxdata.com/products/influxdb-enterprise/) is a hardened version of the open source core InfluxDB that includes additional closed source features designed for production workloads, featuring high availability and horizontal scaling. InfluxDB Enterprise features require a InfluxDB Enterprise license.
+
+#### Configuration
+To enable InfluxDB Enterprise, set the following keys and values in a values file provided to Helm.
```
Comment (naseemkullah) [inline stable/influxdb/README.md:72]: Feel free to provide a sample `enterprise-values.yaml` file as well!

**Review by naseemkullah — COMMENTED** [review]
Don't know much about enterprise, but I left a few comments if you wish to consider them.

File: stable/influxdb/templates/meta-statefulset.yaml (hunk 2 above)
Comment (wgj) [reply stable/influxdb/templates/meta-statefulset.yaml:51]: Sorry, "distinguished" wasn't a great fit here. I meant like a FQDN. InfluxDB Enterprise needs names to be resolvable.

File: stable/influxdb/templates/statefulset.yaml (hunk 1 above)
Comment (wgj) [reply stable/influxdb/templates/statefulset.yaml:92]: Sorry, "distinguished" wasn't a great fit here. I meant like a FQDN. InfluxDB Enterprise needs names to be resolvable.

File: stable/influxdb/values.yaml (hunk 3 above)
Comment (wgj) [reply stable/influxdb/values.yaml:63]: That's a good idea. I'd like to leave as an improvement for a later PR, if that's okay with you.

File: stable/influxdb/README.md (hunk 5 above)
Comment (wgj) [reply stable/influxdb/README.md:72]: Is that preferred? I tend to lean away from multiple values files, as the ordering is non-intuitive and easy to break (and I'm unaware if merging `.env` will do what I want). 

Everything in `.enterprise` could be kept separate, but seeing as how that part is a neat and contained, I think it makes sense to keep colocated.

File: stable/influxdb/templates/meta-statefulset.yaml (hunk 4 above)
Comment (wgj) [reply stable/influxdb/templates/meta-statefulset.yaml]: Thanks!

**Review by naseemkullah — APPROVED** [review]

File: stable/influxdb/values.yaml [hunk 6]
```diff
@@ -56,6 +56,29 @@ persistence:
...
+  clusterSize: 4
+  meta:
+    image:
+      tag: meta
```
Comment (Aisuko) [inline stable/influxdb/values.yaml:70]: Is there could add the link about how to make choice for the Influxdb tag?

File: stable/influxdb/values.yaml [hunk 7]
```diff
@@ -56,6 +56,29 @@ persistence:
   accessMode: ReadWriteOnce
   size: 8Gi
 
+## Deploy InfluxDB Enterprise - License required
+##
+enterprise:
```
Comment (Aisuko) [inline stable/influxdb/values.yaml:61]: I believe we could add the link about the information of Influxdb Enterprise.

**Review by Aisuko — COMMENTED** [review]
Thanks for the new feature, there may some comments, please have a look.

**Review by Aisuko — CHANGES_REQUESTED** [review]
I believe for the Statefulset we need to add the `headless` type of service, please see the link below https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations

**Comment by wgj** [comment]
> I believe for the Statefulset we need to add the `headless` type of service

Line 78 of `README.md` instructs users to use a headless service. It sucks I'm making them do it by hand, but the current `.service.type: ClusterIP` in `values.yaml` is open ended, and I wanted to be polite and not make assumptions for anyone.

File: stable/influxdb/values.yaml (hunk 7 above)
Comment (Aisuko) [reply stable/influxdb/values.yaml:61]: Thanks

**Comment by Aisuko** [comment]
I know what's your mean. I have a consider here, Should we need to make the `service` to multiple just like the `service` file in the `stable/rabbitmq`  ?https://github.com/helm/charts/tree/master/stable/rabbitmq/templates   

This may be useful for new users who do not know `statefulset`. That could be let them know the `statefulset` need to create the service with no `ClusterIP` and it just for the pods belongs to the `statfulset` communicating with each other.  Another reason is out-of-box for the chart, If somebody 
 who is deploying that then they can use in one second instead of need to create the service type in `ClsuterIP`, `NodePort`, `Loadbalance` manually.

**Comment by wgj** [comment]
@Aisuko I'm not sure I fully understand your suggestion, but after thinking about what you said, I thought of this...

How about the `.spec.type` in `templates/service.yaml` use a conditional: if `.Values.enterprise.enabled`, then use a headless service, else continue normally? Is that better for users?

**Comment by naseemkullah** [comment]
statefulsets generally require headless services regardless if enterprise is enabled or not, I would make it more generic. i.e. If you have statefulset you have a headless service.

**Comment by wgj** [comment]
Good point @naseemkullah. I believe this is a "bug" introduced from #17619, which I'm imagining it's been harmless thus far since non-Enterprise deployments are a statefulset of 1. Happy to "fix" this in my branch, or leave it to be part of another PR. 

Let me know what you'd guys like me to do.

**Comment by naseemkullah** [comment]
Yes #17619 was a little bit rushed for such an impactful change imho, i am still pinned to a version of the chart prior to it until everything is sorted out.

Although a separate "hotfix" PR would be the correct thing, its up to you guys :)

**Comment by Aisuko** [comment]
I mean there would have two headless services, one for the existing service another one for enterprise.  And you are right, this is the bug for the PR which changed the source type to `statefulset`.  But, these two different issues.  I would fix this in other PR. For this PR, please just set the `headless` service for enterprise `statefulset`.

**Review by Aisuko — CHANGES_REQUESTED** [review]
Maybe there was something wrong. I mean the Statefulset need two services.  For example "meta-service.yaml" and "meta-service-headless.yaml" would be good. The first one above exposes the service for end-user, another one for Influxdb Statefulset.

Just like https://github.com/helm/charts/tree/master/stable/rabbitmq/templates/svc-headless.yaml and 
https://github.com/helm/charts/tree/master/stable/rabbitmq/templates/svc.yaml

The service for Influxdb Statefulset(not the enterprise version) will be named `service.yaml`, `service-headless.yaml` and this will be included in the next PR.

Thanks for your patient, sir.

**Comment by wgj** [comment]
The metadata pods are not exposed to end-users. After running `kubectl exec influxdb-meta-0 influxd-ctl add-meta ...` & `kubectl exec influxdb-meta-0 influxd-ctl add-data ...` from `README.md`, end-users use InfluxDB data pods only. Would you like me to rename the file `meta-service.yaml` to `meta-service-headless.yaml`?

**Comment by Aisuko** [comment]
Well, I have a consider this.  I saw the contents in README.md. The meta pod only use for the internal Influxdb pods. They do not need to expose themself, right?  

* I saw the content that you committed. https://github.com/helm/charts/pull/19625/commits/1a89ea570674c1493549cebab1047a4501c5ce6f

<details>
<summary>The content of service</summary>

```
type: {{ .Values.service.type }}
  {{- if .Values.service.clusterIP }}	
  clusterIP: {{ .Values.service.clusterIP }}	
  {{- end }}	
  {{- if .Values.service.externalIPs }}	
  externalIPs:	
  {{ toYaml .Values.service.externalIPs | indent 4 }}	
  {{- end }}	
  {{- if .Values.service.loadBalancerIP }}	
  loadBalancerIP: {{ .Values.service.loadBalancerIP }}	
  {{- end }}	
  {{- with .Values.service.loadBalancerSourceRanges }}	
  loadBalancerSourceRanges:	
    {{- toYaml . | nindent 4 }}	
  {{- end }}	
  # publishNotReadyAddresses is used for service discovery of meta and data nodes by querying the service's SRV record.
  publishNotReadyAddresses: true
```
</details>

I have a consider If `{{.Values.service.type }}` was set as `ClusterIP` or `NodePort` by end-users.  They only want to expose the Influxdb service outside the cluster.  But the `meta-service` type will be set the same as `Influxdb data` service type. And the `meta-statefulset` does not have `headless` service. How should we solve this?

**Comment by Aisuko** [comment]
@wgj  Thanks for your patient. We had a discussion about `headless`.  And I saw @naseemkullah also approved this. So, we can merge this.

**Pull request merged**